src/api/03-user-experience/09-google-cast.md

```diff
@@ -0,0 +1,247 @@
+```
+
+## Captions and subtitles
+
+Media can contain additional [tracks of different types]. You can use text tracks to provide captions or subitles for your videos.
+
+```js
+const mediaInfo = new chrome.cast.media.MediaInfo(
+  'http://localhost/video.mp4',                // Provide a URL.
+  'video/mp4; codecs="avc1.640032,mp4a.40.2"', // MIME + codecs.
+);
+
+// Instantiate the Track object first.
+const captionsTrack = new chrome.cast.media.Track(
+  'captions-en', // Unique track identifier.
+  chrome.cast.media.TrackType.TEXT,
+);
+
+captionsTrack.trackContentId = 'http://localhost/captions-en.vtt';
+captionsTrack.subtype = chrome.cast.media.TextTrackType.CAPTIONS;
+captionsTrack.name = 'English CC';
+captionsTrack.language = 'en-US'; // RFC 5646 Language Tag.
+captionsTrack.trackContentType = 'text/vtt';
+
+// The `MediaInfo` object may contain several tracks of various types,
+// how and whether additional tracks are exposed in the UI depends
+// on the receiver application implementation.
+mediaInfo.tracks = [ captionsTrack ];
+```
+
+## UX considerations
+
+When the casting session starts, video playback pauses in the sender application and the `<google-cast-launcher>` changes its styling to indicate the session in progress.
+
+However in many cases it is useful to provide additional indication. One of the common patterns is rendering a video overlay along with displaying a name of the device the media is being casted to.
+
+To achieve this, we can listen for `SESSION_STATE_CHANGED` events and add or remove an overlay element when session is started and ended:
+
+```html
+<style>
+  .video-wrapper {
+    position: relative;
+    display: inline-block;
+  }
+  .cast-overlay {
+    position: absolute;
+    display: grid;
+    place-content: center;
+    inset: 0;
+    background: #222;
+    color: white;
+  }
+  .hidden {
+    display: none;
+  }
+</style>
+
+<div class="video-wrapper">
+  <video src="video.mp4"></video>
+  <div class="cast-overlay hidden">
+    Casting<span class="cast-target"></span>
+  </div>
+</div>
+
+<script>
+  // The following code should be defined within
+  // `function initCastApi() { ... }` defined previously.
+  const context = cast.framework.CastContext.getInstance();
+
+  const overlay = document.querySelector('.cast-overlay');
+  const target = document.querySelector('.cast-target');
+
+  // Bind the overlay visibility state to session state changes.
+  context.addEventListener(
+    cast.framework.CastContextEventType.SESSION_STATE_CHANGED,
+    e => {
+      switch (e.sessionState) {
+        case 'SESSION_ENDED':
+          overlay.classList.remove('hidden');
+          break;
+        case 'SESSION_STARTED':
+        case 'SESSION_RESUMED':
+          const session = context.getCurrentSession();
+          const targetName = session.getCastDevice().friendlyName;
+
+          overlay.classList.add('hidden');
+          target.innerText = ` to ${targetName}`;
```

Comment:
Shall we have "Casting to" and only `targetName` in target to avoid i18n issues?

Reply:
Right, that would make sense. But, is `session.getCastDevice().friendlyName` always non-empty?

The current implementation should only append the `to [device name]` string to the overlay when device name is not empty. I guess we could sub in something like `external device` in cases when the `friendlyName` is empty.

Right now, when the `friendlyName` is empty, the overlay displays just `Casting`.

Reply:
I'd assume it's not empty because it is mandatory according to https://developers.google.com/cast/docs/reference/web_sender/chrome.cast.Receiver

Reply:
Good point. Updated in [e4fb7f7](https://github.com/GoogleChrome/kino/pull/182/commits/e4fb7f7a018355634be2fc0ea61870209c859e3c).